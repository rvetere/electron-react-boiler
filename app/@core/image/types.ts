import { ImageSetItem } from '@graphql/types/generated'
import { ReactNode } from 'react'

export interface IImageProps {
  src: string
  alt: string
  loadingSpinner?: ReactNode
  width?: number
  height?: number
  imageSet?: Array<ImageSetItem | null>
  showImageSet?: boolean
  renderDimensionsAttributes?: boolean
  onLoad?: () => void
}

export interface IImageControllerState {
  isVisible: boolean
  width: number
  height: number
}

export interface IImageViewProps {
  isVisible: boolean
  src: string
  width: number
  height: number
  alt: string
  renderDimensionsAttributes: boolean
  onLoad?: () => void
}

export interface IImageControllerRenderProps {
  width: number
  height: number
  isVisible: boolean
  src: string
  onLoad: () => void
  onIntersection: (ev: IntersectionObserverEntry, unobserve: () => void) => void
}

export interface IImageControllerProps {
  width?: number
  height?: number
  src: string
  onLoad?: () => void
  children: (data: IImageControllerRenderProps) => ReactNode
}

export interface IImageListViewProps {
  imageSet: Array<ImageSetItem | null>
  width: number
  height: number
  isVisible: boolean
  renderDimensionsAttributes: boolean
  animationStarted: boolean
  activeImageIndex: number
  startAnimation: () => void
  endAnimation: () => void
  onLoad?: () => void
}

export interface IImageListControllerRenderProps {
  startAnimation: () => void
  endAnimation: () => void
  animationStarted: boolean
  activeImageIndex: number
}

export interface IImageListControllerProps {
  imageSet: Array<ImageSetItem | null>
  children: (data: IImageListControllerRenderProps) => ReactNode
}

export interface IImageListControllerState {
  animationStarted: boolean
  activeImageIndex: number
}
