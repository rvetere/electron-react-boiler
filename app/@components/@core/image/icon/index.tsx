import classNames from 'class-names'

import { IIconProps } from '@core/image/icon/types'
import aidKit from './components/AidKit'
import styles from './styles.css'

type JsxOrNull = JSX.Element | null

/* tslint:disable no-any variable-name */
export const wrapIcon = (Icon: any, { width = 16, height = 16, className }: IIconProps): JsxOrNull => (
  <Icon width={width} height={height} className={classNames(styles.digicon, className)} />
)

export const AidKit: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(aidKit, props)
