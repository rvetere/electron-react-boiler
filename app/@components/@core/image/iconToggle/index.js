import styles from './styles.css'

export const IconToggle = ({ children }) => (
  <span className={styles.wrapper}>
    <span className={styles.static}>{children[0]}</span>
    <span className={styles.hover}>{children[1]}</span>
  </span>
)

export default IconToggle
