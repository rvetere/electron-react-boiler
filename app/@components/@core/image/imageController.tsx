import CONFIG from '@config'
import noop from '@helpers/noop'
import { PureComponent, ReactNode } from 'react'
import { IImageControllerProps, IImageControllerState } from './types'

export const setSrcByDimensions = (src: string, width: number, height: number, checkForHighDensity: boolean = true): string => {
  if (src.indexOf('?') === -1) {
    return src
  }

  let srcParts = decodeURIComponent(src).split('?')
  let host = srcParts[0]
  let queryString = srcParts[1]
  let indexOfOutputFormat = queryString.indexOf('&output-format')

  // the image service from akamai requires an output-format
  if (indexOfOutputFormat === -1) {
    let format = 'progressive-jpeg'
    // Don't try to convert png/gif to progressive-jpeg, doesn't work
    if (host.endsWith('.png')) {
      format = 'png'
    } else if (host.endsWith('.gif')) {
      format = 'gif'
    }

    queryString += `&output-format=${format}`
    indexOfOutputFormat = queryString.indexOf('&output-format')
  }

  const start = queryString.indexOf('|') + 1
  const end = indexOfOutputFormat
  const dimensions = queryString.slice(start, end)
  const isHighDensity = !CONFIG.IS_SERVER && window.devicePixelRatio > 1

  width = checkForHighDensity && isHighDensity ? Math.ceil(width * 2) : Math.ceil(width)
  height = checkForHighDensity && isHighDensity ? Math.ceil(height * 2) : Math.ceil(height)

  queryString = queryString.replace(dimensions, `${width}:${height}`)

  return encodeURI(`${host}?${queryString}`)
}

export class ImageController extends PureComponent<IImageControllerProps, IImageControllerState> {
  public readonly state: IImageControllerState = {
    isVisible: false,
    width: 300,
    height: 300
  }

  public render(): ReactNode {
    const { children, src } = this.props
    const { isVisible, width, height } = this.state
    const optimizedSrc = setSrcByDimensions(src, width, height, true)

    return children({
      onLoad: this.onLoad,
      onIntersection: this.onIntersection,
      isVisible,
      src: optimizedSrc,
      width,
      height
    })
  }

  private onLoad = (): void => {
    const { onLoad = noop } = this.props
    onLoad()
  }

  private onIntersection = (ev: IntersectionObserverEntry, unobserve: (state: IImageControllerState) => void): void => {
    if (ev.isIntersecting && !this.state.isVisible) {
      let width
      let height

      if (this.props.width && this.props.height) {
        width = this.props.width
        height = this.props.height
      } else if (ev.target && ev.target.parentNode) {
        const parentNode = ev.target.parentNode as Element
        const parentDimensions = parentNode
          ? parentNode.getBoundingClientRect()
          : {
              width: 0,
              height: 0
            }

        width = parentDimensions.width
        height = parentDimensions.height
      } else {
        width = this.state.width
        height = this.state.height
      }

      const newState = { isVisible: true, width: Math.ceil(width), height: Math.ceil(height) }

      this.setState(newState)
      unobserve(newState)
    }
  }
}
