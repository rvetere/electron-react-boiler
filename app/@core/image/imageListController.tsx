import { IInterval } from '@helpers/types'
import { clearRequestInterval, requestInterval } from '@helpers/window'
import { PureComponent, ReactNode } from 'react'
import { IImageListControllerProps, IImageListControllerState } from './types'

export class ImageListController extends PureComponent<IImageListControllerProps, IImageListControllerState> {
  public readonly state: IImageListControllerState = {
    animationStarted: false,
    activeImageIndex: 0
  }

  private animationInterval: number = 1600
  private animationIntervalHandle: IInterval

  public render(): ReactNode {
    const { children } = this.props
    const { animationStarted, activeImageIndex } = this.state

    return children({
      animationStarted,
      activeImageIndex,
      startAnimation: this.startAnimation,
      endAnimation: this.endAnimation
    })
  }

  private endAnimation = (): void => {
    clearRequestInterval(this.animationIntervalHandle)
    this.setState({ activeImageIndex: 0, animationStarted: false })
  }

  private stepAnimation = (): void => {
    const nextIndex = this.state.activeImageIndex + 1
    const maxIndex = this.props.imageSet.length - 1

    const activeImageIndex = nextIndex > maxIndex ? 0 : nextIndex

    this.setState({ activeImageIndex })
  }

  private startAnimation = (): void => {
    clearRequestInterval(this.animationIntervalHandle)
    this.stepAnimation()
    this.animationIntervalHandle = requestInterval(this.stepAnimation, this.animationInterval)
    this.setState({ animationStarted: true })
  }
}
