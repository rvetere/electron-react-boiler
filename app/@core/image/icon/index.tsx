import classNames from 'class-names'
import React from 'react'

import { IIconProps } from '@core/image/icon/types'
import checkbox from './components/Checkbox'
import checkboxActive from './components/CheckboxActive'
import chevronDown from './components/ChevronDown'
import chevronLeft from './components/ChevronLeft'
import chevronRight from './components/ChevronRight'
import chevronUp from './components/ChevronUp'
import close from './components/Close'
import eye from './components/Eye'
import eyeClosed from './components/EyeClosed'
import radio from './components/Radio'
import radioActive from './components/RadioActive'

import styles from './styles.css'

type JsxOrNull = JSX.Element | null

/* tslint:disable no-any variable-name */
export const wrapIcon = (Icon: any, { width = 16, height = 16, className }: IIconProps): JsxOrNull => (
  <Icon width={width} height={height} className={classNames(styles.digicon, className)} />
)

export const ChevronDown: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(chevronDown, props)
export const ChevronLeft: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(chevronLeft, props)
export const ChevronRight: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(chevronRight, props)
export const ChevronUp: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(chevronUp, props)
export const Checkbox: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(checkbox, props)
export const CheckboxActive: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(checkboxActive, props)
export const Close: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(close, props)
export const Eye: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(eye, props)
export const EyeClosed: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(eyeClosed, props)
export const Radio: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(radio, props)
export const RadioActive: React.SFC<IIconProps> = (props: IIconProps): JsxOrNull => wrapIcon(radioActive, props)
