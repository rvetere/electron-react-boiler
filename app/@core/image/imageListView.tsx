import { setSrcByDimensions } from '@core/image/imageController'
import { ImageView } from '@core/image/imageView'
import { ImageSetItem } from '@graphql/types/generated'
import isMobileSafari from '@helpers/isMobileSafari'
import classNames from 'classnames'
import { createRef, PureComponent, ReactNode, RefObject } from 'react'
import styles from './imageListView.css'
import { IImageListViewProps } from './types'

export class ImageListView extends PureComponent<IImageListViewProps> {
  public listRef: RefObject<HTMLUListElement> = createRef()
  public animationTriggerElement: Element | null

  public componentDidMount(): void {
    const listElement = this.listRef.current
    const closestElementStr = isMobileSafari() ? 'div' : 'article'
    const animationTriggerElement = listElement && listElement.closest ? listElement.closest(closestElementStr) : null

    if (listElement && animationTriggerElement) {
      const { startAnimation, endAnimation } = this.props

      this.animationTriggerElement = animationTriggerElement

      animationTriggerElement.addEventListener('mouseover', startAnimation)
      animationTriggerElement.addEventListener('touchstart', startAnimation)

      animationTriggerElement.addEventListener('mouseleave', endAnimation)
      animationTriggerElement.addEventListener('touchend', endAnimation)
      animationTriggerElement.addEventListener('touchcancel', endAnimation)
    }
  }

  public componentWillUnmount(): void {
    if (this.animationTriggerElement) {
      const { startAnimation, endAnimation } = this.props

      this.animationTriggerElement.removeEventListener('mouseover', startAnimation)
      this.animationTriggerElement.removeEventListener('touchstart', startAnimation)

      this.animationTriggerElement.removeEventListener('mouseleave', endAnimation)
      this.animationTriggerElement.removeEventListener('touchend', endAnimation)
      this.animationTriggerElement.removeEventListener('touchcancel', endAnimation)
    }
  }

  public render(): ReactNode {
    const { onLoad, animationStarted, activeImageIndex, imageSet = [], isVisible, width, height, renderDimensionsAttributes } = this.props

    if (!imageSet.length) {
      return null
    }

    return (
      <ul ref={this.listRef} className={styles.imageList}>
        {imageSet.map((imageSetItem: ImageSetItem, index: number) => (
          <li
            key={index}
            className={classNames({
              [styles.showImage]: index === activeImageIndex
            })}
          >
            {(animationStarted || index === 0) && (
              <ImageView
                onLoad={onLoad}
                isVisible={isVisible}
                src={setSrcByDimensions(imageSetItem.source, width, height, true)}
                width={width}
                height={height}
                alt={imageSetItem.alternateText || ''}
                renderDimensionsAttributes={renderDimensionsAttributes}
              />
            )}
          </li>
        ))}
      </ul>
    )
  }
}
