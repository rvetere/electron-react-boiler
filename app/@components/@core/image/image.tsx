import LoadingCircle from '@core/loading/circle'
import Observer from '@researchgate/react-intersection-observer'
import React, { PureComponent, ReactNode } from 'react'
import { ImageController } from './imageController'
import { ImageListController } from './imageListController'
import { ImageListView } from './imageListView'
import { ImageView } from './imageView'
import { IImageControllerRenderProps, IImageListControllerRenderProps, IImageProps } from './types'

export class Image extends PureComponent<IImageProps> {
  public render(): ReactNode {
    const {
      alt,
      src,
      imageSet,
      onLoad,
      width,
      height,
      loadingSpinner = <LoadingCircle centered />,
      showImageSet = true,
      renderDimensionsAttributes = false
    } = this.props

    return (
      <ImageController src={src} onLoad={onLoad} width={width} height={height}>
        {(data: IImageControllerRenderProps): ReactNode => (
          <Observer onChange={data.onIntersection}>
            <span>
              {!data.isVisible && loadingSpinner}

              {showImageSet && imageSet && imageSet.length > 1 ? (
                <ImageListController imageSet={imageSet}>
                  {({ startAnimation, activeImageIndex, animationStarted, endAnimation }: IImageListControllerRenderProps): ReactNode => (
                    <ImageListView
                      animationStarted={animationStarted}
                      activeImageIndex={activeImageIndex}
                      startAnimation={startAnimation}
                      endAnimation={endAnimation}
                      onLoad={data.onLoad}
                      renderDimensionsAttributes={renderDimensionsAttributes}
                      width={data.width}
                      height={data.height}
                      isVisible={data.isVisible}
                      imageSet={imageSet}
                    />
                  )}
                </ImageListController>
              ) : (
                <ImageView
                  onLoad={data.onLoad}
                  isVisible={data.isVisible}
                  src={data.src}
                  width={data.width}
                  height={data.height}
                  alt={alt}
                  renderDimensionsAttributes={renderDimensionsAttributes}
                />
              )}
            </span>
          </Observer>
        )}
      </ImageController>
    )
  }
}
