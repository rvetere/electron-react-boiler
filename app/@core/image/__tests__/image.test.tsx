import { cleanup, render } from 'react-testing-library'
import { Image } from '../image'
import { getImageMock } from './__mocks__/Image.mock'

describe('Image', () => {
  afterEach(cleanup)

  test('Image renders correctly', () => {
    const imageMock = getImageMock()
    const { container } = render(<Image {...imageMock} />)

    expect(container).toMatchSnapshot()
  })
})
