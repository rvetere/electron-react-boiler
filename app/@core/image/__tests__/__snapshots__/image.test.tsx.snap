// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Image renders correctly 1`] = `
<div>
  <observer>
    <span>
      <span
        class="loader spinner centered"
        style="width: 16px; height: 16px;"
      />
      <img
        alt="AltText"
        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAIAAAABCAQAAABeK7cBAAAADUlEQVQIHWP4z/CfAQAF/wH/t3haYQAAAABJRU5ErkJggg=="
      />
      <noscript>
        &lt;img src="https://static.digitecgalaxus.ch/Files/1/3/7/6/2/9/7/8/MSSurface_Jupiter.jpg?fit=inside%7C300:300&output-format=progressive-jpeg" alt="AltText" /&gt;
      </noscript>
    </span>
  </observer>
</div>
`;
